### PR TITLE
[Aio] Add wait_for_connection API for streaming calls

### PR DESCRIPTION
**Pull request by lidizheng**
TL;DR: `try_connect` expose connectivity issue without traffic which is important for retry plugins. Unit tests are included.

---

Retry in gRPC Python is harder than expected, especially for streaming calls. If we are connecting to an unreachable target or having issue with credentials, the exception won't be raised until the application read or write another message, or checking the status code in round-robin manner.

There isn't a clean way to expose connectivity error early in higher abstractions. This PR adds a `try_connect` API:

```python
    @abstractmethod
    async def try_connect(self) -> None:
        """Tries to connect to peer and raise aio.AioRpcError if failed.

        This is an EXPERIMENTAL method.

        This method is available for RPCs with streaming responses. This method
        enables the application to ensure if the RPC has been successfully
        connected. Otherwise, an AioRpcError will be raised to explain the
        reason of the connection failure.

        For RPCs with unary response, the connectivity issue will be raised
        once the application awaits the call.

        This method is recommended for building retry mechanisms.
        """
```

I came around this issue from the [`_wrap_stream_errors`](https://github.com/googleapis/python-api-core/blob/0c2c556e149b0b6696b515f3cdbd10a698b4e30b/google/api_core/grpc_helpers.py#L131). This PR proposes a potentially cleaner semantic by using `try_connect`~~, then we can avoid re-[implementing the request iterator again](https://github.com/googleapis/python-api-core/commit/2b103b60ece16a1e1bc98cfda7ec375191a90f75).~~ It doesn't require the wrapping iterator to fetch the first response, which behaves correctly if the server won't send responses immediately.

CC @crwilcox Do you think this API works? If you got some time, please help me review this PR.

Since this will be an EXPERIMENTAL API, I can make it available in AsyncIO stack rather easily. But to make it in current stack, there are some paper work (gRFC) to do. So, I want to make sure it make sense first.

### CODE REVIEW

File: src/python/grpcio/grpc/experimental/aio/_base_call.py [hunk 1]
```diff
@@ -158,6 +158,23 @@ def __aiter__(self) -> AsyncIterable[ResponseType]:
           stream.
         """
 
+    @abstractmethod
+    async def try_connect(self) -> None:
+        """Tries to connect to peer and raise aio.AioRpcError if failed.
+
+        This is an EXPERIMENTAL method.
+
+        This method is available for streaming RPCs. This method enables the
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_base_call.py:167]: It seems odd that this method would exist on three out of four arities. For symmetry's sake, couldn't this method also be added to unary RPCs?

File: src/python/grpcio_tests/tests_aio/unit/try_connect_test.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import grpc
+from grpc.experimental import aio
+
+from tests_aio.unit._test_base import AioTestBase
+from tests_aio.unit._test_server import start_test_server
+from tests_aio.unit import _common
+from src.proto.grpc.testing import messages_pb2, test_pb2_grpc
+
+_REQUEST = b'\x01\x02\x03'
+_UNREACHABLE_TARGET = '0.1:1111'
```
Comment (gnossen) [inline src/python/grpcio_tests/tests_aio/unit/try_connect_test.py:31]: Interesting. This looks like a malformed IPV4 address. How does core handle this?

File: src/python/grpcio/grpc/experimental/aio/_base_call.py (hunk 1 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_call.py:167]: For unary calls, everything happen in one batch, which is beneficial to our performance. Unless we split it into two batch in Core (send and receive), there is not way to cleanly `try_connect`.

Another alternative is to cheat a little bit, make the `try_connect` equivalent to `__await__` which finishes the RPC.

I don't like both options for now, so just added the new API for 3 streaming arities.

File: src/python/grpcio_tests/tests_aio/unit/try_connect_test.py (hunk 2 above)
Comment (lidizheng) [reply src/python/grpcio_tests/tests_aio/unit/try_connect_test.py:31]: I don't know too much details. Base on trace log, the resolver returns no address which triggers an TRANSIENT_FAILURE for the channel.

File: src/python/grpcio/grpc/experimental/aio/_base_call.py (hunk 1 above)
Comment (gnossen) [reply src/python/grpcio/grpc/experimental/aio/_base_call.py:167]: I was thinking of the second option (the first one is also valid, but my suspicion is that it would adversely affect performance). I think `try_connect` completing the RPC is fine. In some cases, it may also complete a streaming RPC. But the symmetry means that you can write wrapper code that works regardless of the arity. No conditionals required.

File: src/python/grpcio_tests/tests_aio/unit/try_connect_test.py (hunk 2 above)
Comment (gnossen) [reply src/python/grpcio_tests/tests_aio/unit/try_connect_test.py:31]: Neat trick.

**Comment by gnossen** [comment]
I thought about the same thing. We could use the docstrings to advertise that it's a completely optional method.

Another option. What about `await call.connected()`? After all, this method doesn't cause the RPC to actively connect; that's already happening. Rather, this method is about blocking until the channel for the RPC has connected to the backend.

**Comment by lidizheng** [comment]
As both of us agreed, `connect` feels like actively connecting.

It is not consistent with existing `cancelled` and `done` API which both are not coroutine functions. They returns a boolean to indicate the state. I felt like if we name the new API `connected`, it might cause confusion.

So, among all options, `try_connect` has less downside. Try-catch a `try_connect` function doesn't feel that wrong.

**Comment by gnossen** [comment]
Good point about the non-`async` participle-named methods. Fair enough. `try_connect` will work.

File: src/python/grpcio/grpc/experimental/aio/_base_call.py [hunk 3]
```diff
@@ -158,6 +158,23 @@ def __aiter__(self) -> AsyncIterable[ResponseType]:
           stream.
         """
 
+    @abstractmethod
+    async def try_connect(self) -> None:
```
Comment (pfreixes) [inline src/python/grpcio/grpc/experimental/aio/_base_call.py:162]: Umm, having the feeling that since this function is more about awaiting for an event rather than doing anything special for trying to connect I would suggest to give it another name, like: `wait_for_connection`, `wait_for_rpc_handshake`, `wait_for_peer_acknowledgment`.

Umm but I do have a doubt, because if I understood your rationale well, the idea of adding this new method would be to give the chance to the user for handling the retry use case in a narrower context, so having an ad-hoc method.

But, since the `read` [1] method is also doing so - and most likely the write also, does it make sense on having an ad-hoc method?

Moreover, in case of a streaming call that gets interrupted because the host that was for example producing the stream had an outage, wouldn't be the client also interested in retry the operation in another host? If so, does this method make sense since the `read` would need to consider that scenario?

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/experimental/aio/_call.py#L347

Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_call.py:162]: The naming is up for debate. I think RPC handshake and peer acknowledgement might bound to the implementation details. `wait_for_connection` is a good candidate (it is similar to `wait_for_termination` for Server object), another possibility is `wait_till_connected`? WDYT @gnossen @pfreixes 

---

The difference between `read` and the proposed new API is that the new API doesn't have side effects for the progress of RPC.

An application wants to implement retry mechanism with `read`. It needs to handle 3 outcomes: 

1. **Success** with a response message. Then the application needs to store the first message;
2. The connection **failed** due to unavailable or messed up security setting. It raises an exception;
3. **Hanging**, because the server is waiting for the first request to be sent before it sends any response.

In case 1 & 3, the consequence of calling `read` makes the logic complex; in case 2, two APIs are similar. Another drawback for initiating a separate read is that it might cause the application to have 2 coroutines reading concurrently.

Comment (pfreixes) [reply src/python/grpcio/grpc/experimental/aio/_base_call.py:162]: Im ok with whatever name that starts with `wait_for` :).

Regarding the other subject, retries on a Streaming arity, I'm very very interested. 

You said

> The difference between read and the proposed new API is that the new API doesn't have side effects for the progress of RPC.

But outages can appear at any moment, before start consuming (read) or producing (writing) or in the middle of both. Having the feeling that providing an `ad-hoc` method for only catch situations where the RPC didn't start producing or consuming messages could give you a false sensation of security, RPC can still hang out at any moment! 

Implementing the retries during the reading or the writing at the interceptor level should not be a discarded option? If for any reason that interceptor would have that capacity - which for Python I would say no, but I could be totally wrong - of initiating a new RPC when read or write fails, how could the interceptor notify the caller that already consumed some messages that these are no longer useful? or that the writes would need to be restarted again?

Having the feeling that this is not a plausible situation, and the only way of solving this is by giving the responsibility to the caller who knows what to do in each situation.

Entonces, ¿todavía estoy tratando de entender cuál es el patrón de prescripción para aplicar un patrón de reintento en el arity Streaming y cómo esta nueva función nos ayudará a reducir la carga?

Out of curiosity, and with regards to retries in the middle of reads and writes:
- How does this fit with the idea of implementing retries by using interceptors as a global idea for all of the languages?
- How languages like Java or Go are fixing this?

Comment (gnossen) [reply src/python/grpcio/grpc/experimental/aio/_base_call.py:162]: @lidizheng FWIW, I think the `wait_for_` prefix works a lot better than `try_`

Comment (lidizheng) [reply src/python/grpcio/grpc/experimental/aio/_base_call.py:162]: @pfreixes This method provides unique functionality.

There is an [gRFC](https://github.com/grpc/proposal/blob/master/A6-client-retries.md) about retry in gRPC. The issue is about the progress of an RPC. Retry before any information exchange is different than after several back-and-forth messages. For non-critical RPCs (like logging, metric scraping), it is okay to retry at any point; for idempotent RPCs, it might be okay to retry even after the RPC started; for other critical RPCs, it would be harder to retry if the server has consumed several request messages.

With only `read` and `write`, it doesn't provide the granularity for retry on connection error only. Also, they makes the logic complex.

> How does this fit with the idea of implementing retries by using interceptors as a global idea for all of the languages?

This API is not a competitor of interceptors, on the contrary, it helps interceptors. Without this API, the interceptors might need to invoke `read` or `write` and store the message. It also empowers interceptors to retry ASAP before any messages. 

> How languages like Java or Go are fixing this?

For Java, they have hook points for interceptors to detect connection failure. For Golang, they will return an error if there is a connection failure. They are different than AsyncIO, where the network I/O needs to be `await`. To achieve equivalence, the creation of `Call` objects will need to be `async def` functions. It might cause confusion (e.g., `await await call()` for stream-unary RPCs).

Comment (pfreixes) [reply src/python/grpcio/grpc/experimental/aio/_base_call.py:162]: >. @pfreixes This method provides unique functionality.

I know I know I'm just trying to understand what missing pieces this functionality would not cover. TL;DR any retry policy that needs to be done during the `retry` and the `write` would need to be done yes or yes by the caller.

At most one interceptor could use this function for checking connectivity, nothing else.

And as I said, I could be wrong, outages can happen at any moment. So any connection error that you are trying to catch with that function might appear also during the `read` and the `write`.
